File: Haskell/quartic_tests.hs
import Quartic

import Data.Complex
import Test.QuickCheck (quickCheck)

toComplex :: RealFloat a => a -> Complex a
toComplex = (:+ 0)

checks :: [Complex Double] -> Bool
checks vals =
    checkPoly vals && all (elemC vals) roots &&
    -- For Cubic equations we do not find all roots. See comment in Quartic.hs
    (length vals == 3 || all (elemC roots) vals)
    where
        poly = polyFromRoots vals
        roots = solvePoly poly
        elemC ys x = any ((< 0.01) . magnitude . (x -)) ys

checkPoly :: [Complex Double] -> Bool
checkPoly poly =
    checkRoots $ solvePoly poly
    where
        checkRoots [] = all (== 0) poly
        checkRoots roots = all ((< 0.01) . magnitude . evaluatePoly poly) roots

polyFromRoots :: Num a => [a] -> [a]
polyFromRoots = foldl convolve [1] . map (: [-1])

convolve :: Num a => [a] -> [a] -> [a]
convolve [] _ = []
convolve (x:xs) ys = zipWithDefault 0 (+) (map (* x) ys) $ 0 : convolve xs ys

zipWithDefault :: a -> (a -> a -> b) -> [a] -> [a] -> [b]
zipWithDefault _ _ [] [] = []
zipWithDefault d f xs ys =
    f (mhead xs) (mhead ys) : zipWithDefault d f (drop 1 xs) (drop 1 ys)
    where
        mhead [] = d
        mhead (x:_) = x

propsRealQuadratic :: Double -> Double -> Double -> Bool
propsRealQuadratic a b c =
    checkRoots $ solvePoly poly
    where
        poly = map toComplex [a, b, c]
        checkRoots [] =
            -- No roots only for a zero polynom. (or infinitly many roots..)
            poly == [0, 0, 0]
        checkRoots [root] = checkRoots [root, root]
        checkRoots [root0, root1] =
            -- Either both roots are real or they are conjugates.
            (isReal root0 && isReal root1) || root0 == conjugate root1
        checkRoots _ = error "bad resulting number of roots"
        isReal = (== 0) . imagPart

evaluatePoly :: Num a => [a] -> a -> a
evaluatePoly poly x = sum . zipWith (*) poly $ iterate (* x) 1

main :: IO ()
main = do
    quickCheck propsRealQuadratic
    quickCheck $ \a b -> checks [a, b]
    quickCheck $ \a b c -> checks [a, b, c]
    quickCheck $ \a b c d -> checks [a, b, c, d]
    quickCheck $ \a b c d e -> checkPoly [a, b, c, d, e]

File: C/quartic_tests.cpp
#include <rapidcheck.h>

#include <complex>
#include <vector>
#include <algorithm>

#include "quartic.h"

using std::complex;
using std::vector;

// Missing support for std::complex in rapidcheck
// (https://github.com/emil-e/rapidcheck/issues/11)
template <typename T>
struct rc::Arbitrary<complex<T>> {
  static Gen<complex<T>> arbitrary() {
    return gen::map(
      gen::pair(
        Arbitrary<T>::arbitrary(),
        Arbitrary<T>::arbitrary()),
      [](std::pair<T, T> pair) { return complex<T> (pair.first, pair.second); }
      );
  }
};

template<typename T>
T evaluatePoly(const vector<T>& poly, T x)
{
    T y = 0, xE = 1;
    for (T c : poly)
    {
        y += c * xE;
        xE *= x;
    }
    return y;
}

void checkPoly(const vector<complex<double>>& poly)
{
    const int degree = poly.size()-1;
    complex<double> results[degree];
    const int numResults = solve_poly(degree, (const complex_t*) poly.data(), (complex_t*) results);
    if (numResults == 0)
    {
        for (int i = 1; i < poly.size(); ++i)
            RC_ASSERT(poly[i] == 0.0);
        return;
    }
    for (int i = 0; i < numResults; ++i)
        RC_ASSERT(std::abs(evaluatePoly(poly, results[i])) < 0.01);
}

template <typename T>
vector<T> convolve(vector<T> x, vector<T> y)
{
    vector<T> result(x.size()+y.size()-1, 0);
    for (int iX = 0; iX < x.size(); ++iX)
        for (int iY = 0; iY < y.size(); ++iY)
            result[iX+iY] += x[iX] * y[iY];
    return result;
}

template <typename T>
vector<T> polyFromRoots(vector<T> roots)
{
    vector<T> result = { 1 };
    for (T root : roots)
        result = convolve(result, {root, -1});
    return result;
}

void checks(const vector<complex<double>>& vals)
{
    // Avoid too large values due to accuracy issues
    for (const auto& x : vals)
        if (std::norm(x) > 1e20)
            return;

    checkPoly(vals);

    const vector<complex<double>> poly = polyFromRoots(vals);
    const int degree = poly.size()-1;
    complex<double> roots[degree];
    const int numRoots = solve_poly(degree, (const complex_t*) poly.data(), (complex_t*) roots);
    RC_ASSERT(numRoots >= 0);
    const double epsilon = 1e-5;
    for (int iR = 0; iR < numRoots; ++iR)
    {
        const complex<double> root = roots[iR];
        bool found = false;
        for (const auto& v : vals)
            if (std::norm(v - root) < epsilon)
            {
                found = true;
                break;
            }
        RC_ASSERT(found);
    }
    for (const auto& v : vals)
    {
        if (v == 0.0)
            continue;
        bool found = false;
        for (int iR = 0; iR < numRoots; ++iR)
            if (std::norm(v - roots[iR]) < epsilon)
            {
                found = true;
                break;
            }
        RC_ASSERT(found);
    }
}

void propsRealQuadratic(double a, double b, double c)
{
    const vector<complex<double>> poly = {a, b, c};
    complex<double> roots[2];
    const int numRoots = solve_poly(2, (const complex_t*) poly.data(), (complex_t*) roots);
    if (numRoots == 0)
    {
        RC_ASSERT(a == 0.0 && b == 0.0 && c == 0.0);
        return;
    }
    if (numRoots == 1)
    {
        RC_ASSERT(roots[0].imag() == 0.0);
        return;
    }
    RC_ASSERT(numRoots == 2);
    RC_ASSERT(
        (roots[0].imag() == 0.0 && roots[1].imag() == 0.0)
        ||
        (roots[0].real() == roots[1].real() && roots[0].imag() == -roots[1].imag())
        );
}

int main()
{
    typedef complex<double> C;
    rc::check("Real quadratic", propsRealQuadratic);
    rc::check("Degree=1", [](C a) { checks( {a} ); });
    rc::check("Degree=2", [](C a, C b) { checks( {a, b} ); });
    rc::check("Degree=3", [](C a, C b, C c) { checks( {a, b, c} ); });
    rc::check("Degree=4", [](C a, C b, C c, C d) { checks( {a, b, c, d} ); });
    return 0;
}
